File: src/bot/handlers/registration.ts
/** Подключение чата и участники: /setup, /join, /members, /remind, /link. */

import "server-only";

import * as repo from "@/db/repo";
import type { User } from "@/db/schema";
import { t } from "@/i18n";
import {
  TelegramError,
  answerCallbackQuery,
  isGroup,
  keyboard,
  sendMessage,
  type InlineKeyboardMarkup,
  type TgCallbackQuery,
  type TgChat,
  type TgChatMemberUpdated,
  type TgMessage,
} from "../api";
import { isChatAdmin, mentionList, resolveLang, syncUser } from "../context";
import { groupPath, joinViaBotButton, openGroupButton, sitePage, webAppButton } from "../site";

/**
 * Кнопки под сообщениями бота в группе. «Заполнить расписание» открывает
 * группу сразу в мини-приложении, если оно у бота есть, а иначе — личку с
 * ботом, которая привязывает человека к чату и уже оттуда открывает сайт.
 */
async function setupKeyboard(chatId: number, lang: string): Promise<InlineKeyboardMarkup> {
  return keyboard([
    [await openGroupButton(lang, chatId)],
    [{ text: t(lang, "btn_who_filled"), callback_data: "members:show" }],
  ]);
}

/**
 * /link — открыть группу на сайте.
 *
 * Ссылка приходит в личку кнопкой Mini App: Telegram сам подтверждает, кто
 * открыл сайт, поэтому ни токенов в ссылке, ни пароля не нужно.
 */
export async function cmdLink(message: TgMessage): Promise<void> {
  const from = message.from;
  if (!from) return;

  const lang = await resolveLang(message.chat, from);
  if (!sitePage()) {
    await sendMessage({ chat_id: message.chat.id, text: t(lang, "web_not_configured") });
    return;
  }

  await syncUser(from);

  let chat: Awaited<ReturnType<typeof repo.getChat>> = null;
  if (isGroup(message.chat)) {
    chat = await repo.upsertChat(message.chat.id, message.chat.title ?? "");
    await repo.addMembership(message.chat.id, from.id);
  } else {
    const chats = await repo.userChats(from.id);
    chat = chats.length === 1 ? chats[0] : null;
  }

  const path = chat ? groupPath({ slug: await repo.ensureSlug(chat) }) : "/";
  const button = webAppButton(t(lang, chat ? "btn_open_group_site" : "btn_open_site"), path);
  const dm = {
    text: chat ? t(lang, "web_link", { chat: chat.title || "—" }) : t(lang, "web_link_any"),
    parse_mode: "HTML" as const,
    reply_markup: button ? keyboard([[button]]) : undefined,
  };

  if (!isGroup(message.chat)) {
    await sendMessage({ chat_id: message.chat.id, ...dm });
    return;
  }

  try {
    await sendMessage({ chat_id: from.id, ...dm });
    await sendMessage({ chat_id: message.chat.id, text: t(lang, "web_link_sent") });
  } catch (error) {
    // Бот не может написать первым, пока человек не нажал /start.
    if (error instanceof TelegramError && error.isForbidden) {
      await sendMessage({
        chat_id: message.chat.id,
        text: t(lang, "web_link_dm_first"),
        reply_markup: keyboard([[await joinViaBotButton(lang, message.chat.id, "btn_open_bot")]]),
      });
      return;
    }
    throw error;
  }
}

export async function cmdSetup(message: TgMessage): Promise<void> {
  const from = message.from;
  if (!from) return;

  if (!isGroup(message.chat)) {
    const lang = await resolveLang(message.chat, from);
    await sendMessage({ chat_id: message.chat.id, text: t(lang, "only_group") });
    return;
  }

  const chat = await repo.upsertChat(message.chat.id, message.chat.title ?? "");
  const lang = chat.lang;
  if (!(await isChatAdmin(message.chat, from))) {
    await sendMessage({ chat_id: message.chat.id, text: t(lang, "only_admin") });
    return;
  }

  await syncUser(from);
  await repo.addMembership(message.chat.id, from.id);

  await sendMessage({
    chat_id: message.chat.id,
    text: t(lang, "setup_done", { chat: message.chat.title ?? "" }),
    parse_mode: "HTML",
    reply_markup: await setupKeyboard(message.chat.id, lang),
  });
}

export async function cmdJoin(message: TgMessage): Promise<void> {
  const from = message.from;
  if (!from) return;

  if (!isGroup(message.chat)) {
    const lang = await resolveLang(message.chat, from);
    await sendMessage({ chat_id: message.chat.id, text: t(lang, "only_group") });
    return;
  }

  const chat = await repo.upsertChat(message.chat.id, message.chat.title ?? "");
  const lang = chat.lang;
  await syncUser(from);
  const isNew = await repo.addMembership(message.chat.id, from.id);

  await sendMessage({
    chat_id: message.chat.id,
    text: t(lang, isNew ? "join_ok" : "join_already", { name: from.first_name ?? "" }),
    parse_mode: "HTML",
    reply_markup: await setupKeyboard(message.chat.id, lang),
  });
}

async function membersText(chatId: number, lang: string): Promise<string> {
  const people = await repo.chatMembers(chatId);
  if (people.length === 0) return t(lang, "members_empty");

  const marked = await repo.filledIds(people.map((person) => person.userId));
  const filled: User[] = [];
  const missing: User[] = [];
  for (const person of people) {
    (marked.has(person.userId) ? filled : missing).push(person);
  }

  let text = t(lang, "members_title");
  text += t(lang, "members_filled", { count: filled.length, names: mentionList(filled) });
  if (missing.length) {
    text += t(lang, "members_missing", { count: missing.length, names: mentionList(missing) });
  }
  return text;
}

export async function cmdMembers(message: TgMessage): Promise<void> {
  if (!isGroup(message.chat)) {
    const lang = await resolveLang(message.chat, message.from);
    await sendMessage({ chat_id: message.chat.id, text: t(lang, "only_group") });
    return;
  }

  const chat = await repo.upsertChat(message.chat.id, message.chat.title ?? "");
  await sendMessage({
    chat_id: message.chat.id,
    text: await membersText(message.chat.id, chat.lang),
    parse_mode: "HTML",
    reply_markup: await setupKeyboard(message.chat.id, chat.lang),
  });
}

export async function onMembersButton(query: TgCallbackQuery): Promise<void> {
  const message = query.message;
  if (!message) return;
  const chat = await repo.getChat(message.chat.id);
  const lang = chat?.lang ?? "ru";

  await answerCallbackQuery({ callback_query_id: query.id });
  await sendMessage({
    chat_id: message.chat.id,
    text: await membersText(message.chat.id, lang),
    parse_mode: "HTML",
    reply_markup: await setupKeyboard(message.chat.id, lang),
  });
}

/** Один агрегированный пинг вместо N сообщений — бережём лимиты Telegram. */
export async function cmdRemind(message: TgMessage): Promise<void> {
  if (!isGroup(message.chat)) return;

  const chat = await repo.upsertChat(message.chat.id, message.chat.title ?? "");
  const lang = chat.lang;
  if (!(await isChatAdmin(message.chat, message.from))) {
    await sendMessage({ chat_id: message.chat.id, text: t(lang, "only_admin") });
    return;
  }

  const people = await repo.chatMembers(message.chat.id);
  const marked = await repo.filledIds(people.map((person) => person.userId));
  const missing = people.filter((person) => !marked.has(person.userId));

  if (missing.length === 0) {
    await sendMessage({ chat_id: message.chat.id, text: t(lang, "remind_nobody") });
    return;
  }

  await sendMessage({
    chat_id: message.chat.id,
    text: t(lang, "remind_text", { names: mentionList(missing) }),
    parse_mode: "HTML",
    reply_markup: await setupKeyboard(message.chat.id, lang),
  });
}

/** Бота добавили в группу — сразу подсказываем, что делать. */
export async function onMyChatMember(
  update: TgChatMemberUpdated,
  botId: number,
): Promise<void> {
  const member = update.new_chat_member;
  if (!member || member.user.id !== botId) return;
  if (member.status !== "member" && member.status !== "administrator") return;

  const tgChat: TgChat = update.chat;
  if (!isGroup(tgChat)) return;

  const chat = await repo.upsertChat(tgChat.id, tgChat.title ?? "");
  try {
    await sendMessage({
      chat_id: tgChat.id,
      text: t(chat.lang, "setup_done", { chat: tgChat.title ?? "" }),
      parse_mode: "HTML",
      reply_markup: await setupKeyboard(tgChat.id, chat.lang),
    });
  } catch (error) {
    if (!(error instanceof TelegramError && error.isForbidden)) throw error;
  }
}

export { setupKeyboard };

File: legacy-python/bot/handlers/registration.py
"""Подключение чата и регистрация участников: /setup, /join, /members, /remind."""

from __future__ import annotations

import logging

from telegram import InlineKeyboardButton, InlineKeyboardMarkup, Update
from telegram.constants import ChatType, ParseMode
from telegram.error import Forbidden
from telegram.ext import CallbackQueryHandler, ChatMemberHandler, CommandHandler, ContextTypes
from telegram.helpers import create_deep_linked_url

from qairu.db import repo
from qairu.i18n import t
from ..utils import is_chat_admin, mention_list, resolve_lang, sync_user

log = logging.getLogger(__name__)


def _setup_keyboard(bot_username: str, chat_id: int, lang: str) -> InlineKeyboardMarkup:
    """Кнопка deep-link: открывает личку и сразу привязывает человека к чату."""
    url = create_deep_linked_url(bot_username, f"c{chat_id}")
    return InlineKeyboardMarkup(
        [
            [InlineKeyboardButton(t(lang, "btn_fill_schedule"), url=url)],
            [InlineKeyboardButton(t(lang, "btn_who_filled"), callback_data="members:show")],
        ]
    )


async def cmd_link(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    """Персональная ссылка на веб-версию этой группы.

    Токен одноразово выдаётся конкретному человеку и открывает ЕГО расписание —
    то же самое, что он заполнил в боте. Поэтому шлём только в личку.
    """
    message = update.effective_message
    tg_chat = update.effective_chat
    tg_user = update.effective_user
    if not message or not tg_chat or not tg_user:
        return

    base = (context.application.bot_data or {}).get("web_base_url", "")
    async with repo.session() as session:
        lang = await resolve_lang(update, session)
        if not base:
            await message.reply_text(t(lang, "web_not_configured"))
            return

        await sync_user(session, tg_user)
        if tg_chat.type in (ChatType.GROUP, ChatType.SUPERGROUP):
            chat = await repo.upsert_chat(session, tg_chat.id, tg_chat.title or "")
            await repo.add_membership(session, tg_chat.id, tg_user.id)
        else:
            chats = await repo.user_chats(session, tg_user.id)
            chat = chats[0] if chats else None
        if chat is None:
            await message.reply_text(t(lang, "avail_no_members"))
            return
        slug = await repo.ensure_slug(session, chat)
        token = await repo.issue_web_session(session, tg_user.id)
        await session.commit()

    url = f"{base}/g/{slug}?t={token}"
    keyboard = InlineKeyboardMarkup([[InlineKeyboardButton(t(lang, "btn_open_web"), url=url)]])

    if tg_chat.type in (ChatType.GROUP, ChatType.SUPERGROUP):
        try:
            await context.bot.send_message(tg_user.id, t(lang, "web_link"),
                                           parse_mode=ParseMode.HTML, reply_markup=keyboard)
            await message.reply_text(t(lang, "web_link_sent"))
        except Forbidden:
            await message.reply_text(
                t(lang, "web_link_dm_first"),
                reply_markup=_setup_keyboard(context.bot.username, tg_chat.id, lang),
            )
    else:
        await message.reply_text(t(lang, "web_link"), parse_mode=ParseMode.HTML,
                                 reply_markup=keyboard)


async def cmd_setup(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    message = update.effective_message
    tg_chat = update.effective_chat
    tg_user = update.effective_user
    if not message or not tg_chat or not tg_user:
        return

    if tg_chat.type not in (ChatType.GROUP, ChatType.SUPERGROUP):
        async with repo.session() as session:
            lang = await resolve_lang(update, session)
        await message.reply_text(t(lang, "only_group"))
        return

    async with repo.session() as session:
        chat = await repo.upsert_chat(session, tg_chat.id, tg_chat.title or "")
        lang = chat.lang
        if not await is_chat_admin(update, context):
            await message.reply_text(t(lang, "only_admin"))
            return
        await sync_user(session, tg_user)
        await repo.add_membership(session, tg_chat.id, tg_user.id)
        await session.commit()

    await message.reply_text(
        t(lang, "setup_done", chat=tg_chat.title or ""),
        parse_mode=ParseMode.HTML,
        reply_markup=_setup_keyboard(context.bot.username, tg_chat.id, lang),
    )


async def cmd_join(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    message = update.effective_message
    tg_chat = update.effective_chat
    tg_user = update.effective_user
    if not message or not tg_chat or not tg_user:
        return

    if tg_chat.type not in (ChatType.GROUP, ChatType.SUPERGROUP):
        async with repo.session() as session:
            lang = await resolve_lang(update, session)
        await message.reply_text(t(lang, "only_group"))
        return

    async with repo.session() as session:
        chat = await repo.upsert_chat(session, tg_chat.id, tg_chat.title or "")
        lang = chat.lang
        await sync_user(session, tg_user)
        is_new = await repo.add_membership(session, tg_chat.id, tg_user.id)
        await session.commit()

    key = "join_ok" if is_new else "join_already"
    await message.reply_text(
        t(lang, key, name=tg_user.first_name),
        parse_mode=ParseMode.HTML,
        reply_markup=_setup_keyboard(context.bot.username, tg_chat.id, lang),
    )


async def cmd_leave(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    """Выйти из списка участников чата — расписание при этом сохраняется."""
    message = update.effective_message
    tg_chat = update.effective_chat
    tg_user = update.effective_user
    if not message or not tg_chat or not tg_user:
        return
    if tg_chat.type not in (ChatType.GROUP, ChatType.SUPERGROUP):
        async with repo.session() as session:
            lang = await resolve_lang(update, session)
        await message.reply_text(t(lang, "only_group"))
        return

    async with repo.session() as session:
        chat = await repo.get_chat(session, tg_chat.id)
        lang = chat.lang if chat else "ru"
        await repo.remove_membership(session, tg_chat.id, tg_user.id)
        await session.commit()

    await message.reply_text(t(lang, "leave_ok", name=tg_user.first_name),
                             parse_mode=ParseMode.HTML)


async def _members_text(session, chat_id: int, lang: str) -> str:
    users = await repo.chat_members(session, chat_id)
    if not users:
        return t(lang, "members_empty")

    filled, missing = [], []
    for user in users:
        (filled if await repo.is_filled(session, user.user_id) else missing).append(user)

    text = t(lang, "members_title")
    text += t(lang, "members_filled", count=len(filled), names=mention_list(filled))
    if missing:
        text += t(lang, "members_missing", count=len(missing), names=mention_list(missing))
    return text


async def cmd_members(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    message = update.effective_message
    tg_chat = update.effective_chat
    if not message or not tg_chat:
        return
    if tg_chat.type not in (ChatType.GROUP, ChatType.SUPERGROUP):
        async with repo.session() as session:
            lang = await resolve_lang(update, session)
        await message.reply_text(t(lang, "only_group"))
        return

    async with repo.session() as session:
        chat = await repo.upsert_chat(session, tg_chat.id, tg_chat.title or "")
        lang = chat.lang
        text = await _members_text(session, tg_chat.id, lang)
        await session.commit()

    await message.reply_text(
        text,
        parse_mode=ParseMode.HTML,
        reply_markup=_setup_keyboard(context.bot.username, tg_chat.id, lang),
    )


async def on_members_button(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    query = update.callback_query
    tg_chat = update.effective_chat
    if not query or not tg_chat:
        return
    async with repo.session() as session:
        chat = await repo.get_chat(session, tg_chat.id)
        lang = chat.lang if chat else "ru"
        text = await _members_text(session, tg_chat.id, lang)
    await query.answer()
    await context.bot.send_message(
        tg_chat.id,
        text,
        parse_mode=ParseMode.HTML,
        reply_markup=_setup_keyboard(context.bot.username, tg_chat.id, lang),
    )


async def cmd_remind(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    """Один агрегированный пинг вместо N сообщений — бережём лимиты Telegram."""
    message = update.effective_message
    tg_chat = update.effective_chat
    if not message or not tg_chat:
        return
    if tg_chat.type not in (ChatType.GROUP, ChatType.SUPERGROUP):
        return

    async with repo.session() as session:
        chat = await repo.upsert_chat(session, tg_chat.id, tg_chat.title or "")
        lang = chat.lang
        if not await is_chat_admin(update, context):
            await message.reply_text(t(lang, "only_admin"))
            return
        users = await repo.chat_members(session, tg_chat.id)
        missing = [u for u in users if not await repo.is_filled(session, u.user_id)]
        await session.commit()

    if not missing:
        await message.reply_text(t(lang, "remind_nobody"))
        return

    await message.reply_text(
        t(lang, "remind_text", names=mention_list(missing)),
        parse_mode=ParseMode.HTML,
        reply_markup=_setup_keyboard(context.bot.username, tg_chat.id, lang),
    )


async def on_my_chat_member(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    """Бота добавили в группу — сразу подсказываем, что делать."""
    result = update.my_chat_member
    if not result or not result.new_chat_member:
        return
    if result.new_chat_member.user.id != context.bot.id:
        return
    if result.new_chat_member.status not in ("member", "administrator"):
        return
    tg_chat = result.chat
    if tg_chat.type not in (ChatType.GROUP, ChatType.SUPERGROUP):
        return

    async with repo.session() as session:
        chat = await repo.upsert_chat(session, tg_chat.id, tg_chat.title or "")
        lang = chat.lang
        await session.commit()

    try:
        await context.bot.send_message(
            tg_chat.id,
            t(lang, "setup_done", chat=tg_chat.title or ""),
            parse_mode=ParseMode.HTML,
            reply_markup=_setup_keyboard(context.bot.username, tg_chat.id, lang),
        )
    except Forbidden:
        log.warning("Нет прав писать в чат %s", tg_chat.id)


def register(app) -> None:
    app.add_handler(CommandHandler("setup", cmd_setup))
    app.add_handler(CommandHandler("join", cmd_join))
    app.add_handler(CommandHandler("members", cmd_members))
    app.add_handler(CommandHandler("leave", cmd_leave))
    app.add_handler(CommandHandler("link", cmd_link))
    app.add_handler(CommandHandler("web", cmd_link))
    app.add_handler(CommandHandler("remind", cmd_remind))
    app.add_handler(CallbackQueryHandler(on_members_button, pattern=r"^members:show$"))
    app.add_handler(ChatMemberHandler(on_my_chat_member, ChatMemberHandler.MY_CHAT_MEMBER))
